src/mpid/ch4/netmod/ofi/ofi_rma.c

```diff
@@ -396,7 +388,7 @@ int MPIDI_OFI_pack_put(const void *origin_addr, int origin_count,
     req->noncontig.put.origin.total_bytes = origin_bytes;
 
     /* target */
-    req->noncontig.put.target.base = target_base;
+    req->noncontig.put.target.base = (void *) target_mem.addr;
```

Comment:
We cannot. `target_mem.addr` (`target_mr.addr` in the upcoming push) is calculated in `MPIDI_OFI_prepare_target_mr`. We always need an integer type or `char*` for arithmetic calculation. Thus, we end up cast it either in `MPIDI_OFI_prepare_target_mr` or in `MPIDI_OFI_pack_put`.

Reply:
Semantically, both `base` and `addr` is an address, so I think we should typed as `void *` to be clear. On doing arithmetic, I would always like to see an explicit cast (or the compiler will warn us, which is a good thing). It makes sense to see cast there since the context (need for arithmetic) is there. On the other hand, if you avoid the explicit cast at arithmetic sites, then we will always wonder whether you are doing pointer arithmetics correctly.

Reply:
The same field is used at contig put/get as `uint64_t`. Now the question becomes whether we prefer to do casting in noncontig put/get or in contig put/get. I'd keep the integer type so casting happens at only a single place.

https://github.com/pmodels/mpich/blob/402d0ef99c805d81f958d10c6d10d4251424110b/src/mpid/ch4/netmod/ofi/ofi_rma.h#L263-L265

Reply:
You are using the past code to justify the new code -- a wrong logic. It is the same question that why we type `riov.addr` as `uint64_t` but since it's old code, it's a separate issue. Semantically an `addr` should be `void *`. With `uint64_t`, we need question how it behaves when we cast back and forth on a different architecture, such as 32-bit. Having explicit cast where the reason is clear in the context (such as arithmetic) is fine. I am only bothered by the case when it is non-intuitive, such as casting an address to `void *` (automatic conversion that you don't need for any pointers) or not casting when you doing address arithmetic. An explicit type cast ask the reviewer to pause and figure out the reason behind, so if the reason is legitimate,  such as due to exiting/past types, it is understandable.

Reply:
Reconciled. `uint64_t` cast is needed due to `libfabric` API, out of our control. Use of an integer type signifies the nature of remote address.